training/controllers/articles.py

```diff
@@ -0,0 +1,113 @@
+    order_by = request.headers.get('order_by_date', 'DESC')
+    if order_by == "ASC":
+        return Article.query.filter_by(user_id=user.id).order_by(Article.created_at.asc()).all()
+    elif order_by == "DESC":
+        return Article.query.filter_by(user_id=user.id).order_by(Article.created_at.desc()).all()
+    else:
+        return Article.query.filter_by(user_id=user.id).order_by(Article.created_at.desc()).all()
+
+
+def get_user(username):
+    if username is None:
+        user = g.user
+    else:
+        user = User.query.filter_by(username=username).first()
+    return user
+
+
+@bp.route('/list', methods=['GET'])
+@bp.route('/list/<username>', methods=['GET'])
+@login_required
+def list_articles(username=None):
+    user = get_user(username)
+    if user is None:
+        return jsonify({'Detail': 'Bad URL, "user" not found.'}), 404
+
+    MAX_LIMIT = 20
+    response = {
+        'Detail': 'List of articles.'
+    }
+
+    if not request.args:
+        articles_list = get_articles(user)
+        response['Total articles'] = len(articles_list)
+    else:
+        # We get the parameters
+        try:
+            start = request.args.get('start', default=1, type=int)
+            limit = request.args.get('limit', default=MAX_LIMIT, type=int)
+        except ValueError:
+            return jsonify({'Detail': 'Bad URL, check your URL parameters.'}), 400
+
+        # We get all articles from logged user.
+        articles_list_all = get_articles(user)
+
+        # We validate parameters.
+        length = len(articles_list_all)
+
+        if start < 1 or limit < 1:
+            return jsonify({'Detail': 'Bad URL, check your URL parameters (Negative or Zero values).'}), 400
+        if start > length:
+            return jsonify({'Detail': 'Bad URL, check your URL parameters (Start parameter number is too large).'}), 400
+        if limit > MAX_LIMIT:
+            return jsonify({'Detail': f'Bad URL, check your URL parameters (Limit parameter number is too large, '
+                                      f'Max Limit={MAX_LIMIT}).'}), 400
+
+        # We get the sublist of articles based on the parameters.
+        end = start - 1 + limit
+        if length < start-1+limit:
+            end = length
+
+        articles_list = articles_list_all[start-1:end]
+
+        # We add interesting information about the original request to the response body.
+        response['Total articles'] = length
+        response['Start parameter'] = start
+        response['Limit parameter'] = limit
+
+    # We add to the response body the items that were selected.
+    article_schema = ArticleSchema()
+    for article in articles_list:
+        article_schema.dump(article)
+        if 'Articles' in response:
+            output = article_schema.dump(article)
+            response['Articles'].append(output)
+        else:
+            output = article_schema.dump(article)
+            response['Articles'] = [output]
+
+    return jsonify(response), 200
```

Comment:
What if we want to add a new model in the future and return a paginated response? This wouldn't scale at all. We should think of a better way. I'll leave you some tips:

1. I recommend using simpler and standard keys in the response, for example:
```
{ "limit": 20, "offset": 10, "total": 25, "result": [{...article 1}, {...article 2}, ...] }
```
2. The pagination must happen at database level, so you must call `limit()` and `offset()` functions of the sqlalchemy query.
3. You can write a little `paginate()` function to help you with all this, roughly like:
```
def paginate(queryset, limit, offset, schema_klass):
   return {
   	"total": queryset.count(),
   	"result":  list(map(lambda r: schema_klass.dump(r), queryset.limit(limit).offset(offset))),
   	"limit": limit,
   	"offset": offset,
   }
```

Reply:
That was a very good feedback comment, thanks. I will rewrite the list articles function based on your recommendations.